### PR TITLE
Rearrangements

### PR DESCRIPTION
**Pull request by brucemiller**
This PR continues the previous TeX-pool reorg PR in preparation for more extensive refactoring.
* move several helper functions and ligature patterns to more appropriate "Engine" modules
* recognizes more font names
* more accurate initialization of various `\mathcode`, `\sfcode`
* simulate TeX-plain's register allocation better.

### CODE REVIEW

File: lib/LaTeXML/Common/Font.pm [hunk 1]
```diff
@@ -177,16 +186,17 @@ sub decodeFontname {
     if (my $ffam = lookupFontFamily($fam)) { map { $props{$_} = $$ffam{$_} } keys %$ffam; }
     if (my $fser = lookupFontSeries($ser)) { map { $props{$_} = $$fser{$_} } keys %$fser; }
     if (my $fsh  = lookupFontShape($shp))  { map { $props{$_} = $$fsh{$_} } keys %$fsh; }
-    $size = 1 unless $size;    # Yes, also if 0, "" (from regexp)
-    $size = $at if defined $at;
-    $size *= $scaled if defined $scaled;
+    $size        = 1 unless $size;    # Yes, also if 0, "" (from regexp)
+    $size        = $at           if defined $at;
+    $size        = $at * $scaled if defined $scaled;
```
Comment (dginev) [inline lib/LaTeXML/Common/Font.pm:191]: Maybe the 3 size assignments here can be compacted to avoid edge cases:
```perl
$size = (defined $at) ? (defined $scaled ? $at * $scaled : $at) : 1;
```

Sample test:
```perl
sub trysize {
  my ($at, $scaled) = @_;
  my $size = (defined $at) ? (defined $scaled ? $at * $scaled : $at) : 1;
  return $size
}

print trysize(undef, undef),"\n";
print trysize(undef, 0.5),"\n";
print trysize(0.5, undef),"\n";
print trysize(0.5, 0.5),"\n";

# output:
# 1
# 1
# 0.5
# 0.25

```

Comment (dginev) [reply lib/LaTeXML/Common/Font.pm:191]: That said, the new code will not scale a previously available `$size`, it will only scale with respect to the `$at` value. Is that change intended?

File: lib/LaTeXML/Common/Font.pm [hunk 2]
```diff
@@ -266,6 +276,7 @@ sub stringify {
   no warnings 'recursion';
   my ($self) = @_;
   my ($fam, $ser, $shp, $siz, $col, $bkg, $opa, $enc, $lang, $mstyle, $flags) = @$self;
+  # !!!!!
```
Comment (dginev) [inline lib/LaTeXML/Common/Font.pm:279]: What could be some words to describe that exclamation?

File: lib/LaTeXML/Common/Font.pm [hunk 3]
```diff
@@ -280,6 +291,16 @@ sub stringify {
...
 
+# Return a Fontinfo-like hash
+# Eventually a more integrated representation of Fonts that accommodates
+# both low-level TeX-like commands, and higher-level CSS-like ones.
+sub asFontinfo {
+  my ($self) = @_;
+  my ($fam, $ser, $shp, $siz, $col, $bkg, $opa, $enc, $lang, $mstyle, $flags) = @$self;
+  return { family => $fam, series => $ser, shape => $shp, size => $siz,
+    color    => $col,          background => $bkg,  opacity   => $opa,
+    encoding => $enc || 'OT1', language   => $lang, mathstyle => $mstyle }; }
```
Comment (dginev) [inline lib/LaTeXML/Common/Font.pm:302]: `$flags` are not returned?

File: lib/LaTeXML/Common/Font.pm [hunk 4]
```diff
@@ -328,6 +349,7 @@ sub relativeTo {
   my ($self, $other) = @_;
   my ($fam,  $ser,  $shp,  $siz,  $col,  $bkg,  $opa,  $enc,  $lang,  $mstyle,  $flags)  = @$self;
   my ($ofam, $oser, $oshp, $osiz, $ocol, $obkg, $oopa, $oenc, $olang, $omstyle, $oflags) = @$other;
+  # !!!!
```
Comment (dginev) [inline lib/LaTeXML/Common/Font.pm:352]: [2] What could be some words to describe that exclamation?

File: lib/LaTeXML/Common/Font/Metric.pm [hunk 5]
```diff
@@ -152,12 +153,15 @@ sub process_lig_kern {
       $prognum = 256 * $op + $remainder; $firstinstr = 0;
       next; }
     $next = $$fontmap[$next];
+    $next = $$next[0] if ref $next;
```
Comment (dginev) [inline lib/LaTeXML/Common/Font/Metric.pm:156]: The changes handling dual type between arrayref vs scalar happen a bit often in this subroutine. This will be somewhat painful to port to Rust, but also makes me wonder why both are allowed in the first place.

Something that will be on my plate a little later, jsut noting it here.

File: lib/LaTeXML/Core/Definition/CharDef.pm [hunk 6]
```diff
@@ -47,13 +47,17 @@ sub invoke {
   my $mathglyph = $$self{mathglyph};
   # A dilemma: If the \chardef were in a style file, you're prefer to revert to the $cs
   # but if defined in the document source, better to use \char ###\relax, so it still "works"
-  if (defined $mathglyph) {    # Must be a math char
+  my $src   = $$self{locator} && $$self{locator}->toString;
+  my $local = $src && $src !~ /\.(?:sty|ltxml|ltxmlc)/;    # Dumps currently have undefined src!
```
Comment (dginev) [inline lib/LaTeXML/Core/Definition/CharDef.pm:51]: Is `.ltxmlc` a new extension? We don't have them in the repository at the moment. If it's dump related, why `c` and not `d`? Or even better `.ltxml.dump` or `.dump.ltxml`

File: lib/LaTeXML/Engine/LaTeX.pool.ltxml [hunk 7]
```diff
@@ -4889,6 +4889,7 @@ DefConstructor('\@framebox[Dimension][]{}',
           $document->setAttribute($c[0], $k => $v); } } } }
 );
 
+AssignValue(allocated_boxes => 0);
```
Comment (dginev) [inline lib/LaTeXML/Engine/LaTeX.pool.ltxml:4892]: Maybe the AssignValue should be `'global'` scoped? Likely identical, until it isn't :>

File: lib/LaTeXML/Engine/TeX_Fonts.pool.ltxml [hunk 8]
```diff
@@ -285,5 +285,38 @@ DeclareFontMap('OMX',
...
+# These ligatures are also handled by TeX.
+# However, it appears that decent modern fonts in modern browsers handle these at that level.
+# So it's likely not worth doing it at the conversion level, possibly adversely affecting search.
+# DefLigature(qr{ff},               "\x{FB00}", fontTest => \&nonTypewriterT1);
+# DefLigature(qr{fi},               "\x{FB01}", fontTest => \&nonTypewriterT1);
+# DefLigature(qr{fl},               "\x{FB02}", fontTest => \&nonTypewriterT1);
+# DefLigature(qr{ffi},              "\x{FB03}", fontTest => \&nonTypewriterT1);
+# DefLigature(qr{ffl},              "\x{FB04}", fontTest => \&nonTypewriterT1);
+
+DefLigature(qr{\.\.\.}, "\x{2026}", fontTest => sub { $_[0]->getFamily ne 'typewriter'; });  # ldots
```
Comment (dginev) [inline lib/LaTeXML/Engine/TeX_Fonts.pool.ltxml:319]: this test can also be `\&nonTypewriter`

File: lib/LaTeXML/Engine/TeX_Paragraph.pool.ltxml [hunk 9]
```diff
@@ -68,6 +68,17 @@ DefConstructorI('\noindent', undef, sub {
...
 
+sub alignLine {
+  my ($document, $line, $alignment) = @_;
+  if ($document->isOpenable('ltx:p')) {
+    $document->insertElement('ltx:p', $line, class => 'ltx_align_' . $alignment); }
+  elsif ($document->isOpenable('ltx:text')) {
+    $document->insertElement('ltx:text', $line, class => 'ltx_align_' . $alignment);
+    $document->insertElement('ltx:break'); }
+  else {
+    $document->absorb($line); }
```
Comment (dginev) [inline lib/LaTeXML/Engine/TeX_Paragraph.pool.ltxml:79]: should this case emit an Info or Warning message? Silently absorbing while dropping the alignment may be worth logging.

File: lib/LaTeXML/Engine/plain.pool.ltxml [hunk 10]
```diff
@@ -237,36 +207,20 @@ DefPrimitive('\wlog{}', sub {
...
+\outer\def\newmuskip{\alloc@3\muskip\muskipdef\@cclv}
+\outer\def\newbox{\alloc@4\box\chardef\insc@unt}
+%%\let\newtoks=\relax % we do this to allow plain.tex to be read in twice
+\outer\def\newhelp#1#2{\newtoks#1#1\expandafter{\csname#2\endcsname}}
+\outer\def\newtoks{\alloc@5\toks\toksdef\@cclv}
+\outer\def\newread{\alloc@6\read\chardef\sixt@@n}
+\outer\def\newwrite{\alloc@7\write\chardef\sixt@@n}
+\outer\def\newfam{\alloc@8\fam\chardef\sixt@@n}
+\outer\def\newlanguage{\alloc@9\language\chardef\@cclvi}
+EoTeX
```
Comment (dginev) [inline lib/LaTeXML/Engine/plain.pool.ltxml:223]: Dropping Perl definitions in favor of RawTeX? Scary. Hopefully more reliable, but something to keep an eye on while testing...

**Review by dginev — COMMENTED** [review]
Nice progress!

I left comments on the pieces I understand already, and you have my "fingers crossed" for the TeX-native transitional changes I only understand in spirit.

File: lib/LaTeXML/Common/Font.pm (hunk 1 above)
Comment (brucemiller) [reply lib/LaTeXML/Common/Font.pm:191]: Essentially a typo, from backing off a change that wasn't ready. The 2nd line should indeed have been ```$size=$size * $scaled if $scaled;``.  Probably those lines will disappear, or be handled differently in the next PR, so not worth optimizing. commit on the way...

File: lib/LaTeXML/Common/Font.pm (hunk 2 above)
Comment (brucemiller) [reply lib/LaTeXML/Common/Font.pm:279]: that got added? Hmm...

File: lib/LaTeXML/Common/Font/Metric.pm (hunk 5 above)
Comment (brucemiller) [reply lib/LaTeXML/Common/Font/Metric.pm:156]: Indeed, there's a few such changes that leaked into this PR, but are really about the upcoming one. The idea is for the FontMap to store not only the equivalent Unicode, but some set of properties/attributes (esp. math related ones).  Ie. all still up for discussion.

File: lib/LaTeXML/Core/Definition/CharDef.pm (hunk 6 above)
Comment (brucemiller) [reply lib/LaTeXML/Core/Definition/CharDef.pm:51]: Exactly for dumps and your naming suggestion sounds good

File: lib/LaTeXML/Engine/LaTeX.pool.ltxml (hunk 7 above)
Comment (brucemiller) [reply lib/LaTeXML/Engine/LaTeX.pool.ltxml:4892]: will do.

File: lib/LaTeXML/Engine/TeX_Fonts.pool.ltxml (hunk 8 above)
Comment (brucemiller) [reply lib/LaTeXML/Engine/TeX_Fonts.pool.ltxml:319]: Ah the benefits of good modularity! will do

File: lib/LaTeXML/Engine/TeX_Paragraph.pool.ltxml (hunk 9 above)
Comment (brucemiller) [reply lib/LaTeXML/Engine/TeX_Paragraph.pool.ltxml:81]: not a bad idea...

File: lib/LaTeXML/Engine/plain.pool.ltxml (hunk 10 above)
Comment (brucemiller) [reply lib/LaTeXML/Engine/plain.pool.ltxml:223]: Yeah, scary, but really the underlying theme: most of `plain.pool.ltxml` could disappear if (some incarnation of) `plain.tex` can be read in.

**Review by dginev — APPROVED** [review]

**Comment by brucemiller** [comment]
Last commit reverts all the FontMap tweaks that were in `LaTeXML::Common::Font::Metric`, since those were too early, and probably won't even happen in the end.
